File: python/evm_state/postgres.py
"""Consistent, fail-closed diagnostics for the legacy PostgreSQL current tables.

These commands do not publish checkpoints. The root check proves captured state
at the exact database head; the sampled RPC check cannot establish completeness.
The legacy snapshot is materialized in memory, so use ClickHouse's streaming
checkpoint verifier for large accounts.
"""
import argparse
import json
import os
import subprocess

from .proof import VerificationError, address, unhex, verify_account, verify_complete, verify_metadata
from .rpc import RPC


def snapshot(selected=None, limit=None, dsn=None):
    # Only normalized, fixed-length hex and validated integers reach SQL. A
    # single statement gives every subquery the same PostgreSQL MVCC snapshot.
    selected = address(selected) if selected is not None else None
    if limit is not None and (isinstance(limit, bool) or not isinstance(limit, int) or limit < 1):
        raise ValueError("storage sample limit must be a positive integer")
    where = f"WHERE a.address='{selected}'" if selected else ""
    storage_where = f"WHERE address='{selected}'" if selected else ""
    bound = f"LIMIT {limit}" if limit is not None else ""
    query = f"""SELECT json_build_object(
        'header', (SELECT row_to_json(h) FROM
            (SELECT block_num AS number,block_hash AS hash,state_root FROM blocks ORDER BY block_num DESC LIMIT 1) h),
        'accounts', (SELECT coalesce(json_agg(row_to_json(a)), '[]'::json) FROM
            (SELECT a.address,a.nonce::text AS nonce,a.balance::text AS balance,a.code_hash,
                '0x' || encode(c.code,'hex') AS code,
                greatest(a.block_num,a.balance_block_num,a.nonce_block_num,a.code_block_num,c.first_block_num) AS last_block
             FROM accounts a LEFT JOIN code c ON a.code_hash=c.code_hash {where} ORDER BY a.address) a),
        'storage', (SELECT coalesce(json_agg(row_to_json(s)), '[]'::json) FROM
            (SELECT address,slot,value,block_num AS last_block FROM storage {storage_where}
             ORDER BY address,slot {bound}) s))"""
    try:
        result = subprocess.run(["psql", "-XAt", "--set=ON_ERROR_STOP=1", "--dbname", dsn or os.environ.get(
            "PG_DSN", "postgresql://dev-node:insecure-change-me-in-prod@localhost:5432/dev-node"), "-c", query],
            capture_output=True, text=True, timeout=120)
    except subprocess.TimeoutExpired:
        # TimeoutExpired includes command arguments, potentially the DSN password.
        raise RuntimeError("PostgreSQL snapshot query timed out") from None
    if result.returncode:
        raise RuntimeError(f"PostgreSQL snapshot query failed (exit {result.returncode}); check the database connection and schema")
    value = json.loads(result.stdout)
    if selected and [row["address"] for row in value["accounts"]] != [selected]:
        raise VerificationError("selected account has no complete metadata row in the database")
    return value


def validate_snapshot(value, requested_block=None):
    header = value.get("header")
    if not header or not value.get("accounts"):
        raise VerificationError("database snapshot has no head or account metadata")
    number = header["number"]
    if isinstance(number, bool) or not isinstance(number, int) or number < 0:
        raise VerificationError("invalid database head")
    if requested_block is not None and requested_block != number:
        raise VerificationError("current tables only support their captured head; --block cannot select historical state")
    unhex(header["hash"], 32)
    unhex(header["state_root"], 32)
    accounts = {}
    for row in value["accounts"]:
        selected = address(row["address"])
        if selected in accounts:
            raise VerificationError("duplicate database account")
        if any(row.get(k) is None for k in ["nonce", "balance", "code_hash", "code"]):
            raise VerificationError("missing required account metadata or bytecode; partial state cannot pass verification")
        try:
            nonce = int(row["nonce"])
            if str(nonce) != str(row["nonce"]) or isinstance(row["nonce"], bool):
                raise ValueError()
        except (ValueError, TypeError) as error:
            raise VerificationError("database nonce must be an exact integer") from error
        if not 0 <= nonce < 2**64:
            raise VerificationError("database nonce exceeds uint64")
        unhex(row["code_hash"], 32)
        unhex(row["code"])
        accounts[selected] = {**row, "nonce": nonce}
    keys = set()
    for row in value["accounts"] + value["storage"]:
        last = row.get("last_block")
        if isinstance(last, bool) or not isinstance(last, int) or not 0 <= last <= number:
            raise VerificationError("database state is newer than its block marker or has invalid provenance")
    for row in value["storage"]:
        if address(row["address"]) not in accounts:
            raise VerificationError("storage account has no verified metadata")
        unhex(row["slot"], 32)
        unhex(row["value"], 32)
        key = (row["address"], row["slot"])
        if key in keys:
            raise VerificationError("duplicate database storage slot")
        keys.add(key)
    return accounts


def verify(value, bundle, complete=False, rpc=None, requested_block=None):
    accounts = validate_snapshot(value, requested_block)
    header = value["header"]
    if bundle["chain_id"] != 56 or any(header[k] != bundle["header"][k] for k in ["number", "hash", "state_root"]):
        raise VerificationError("database head differs from the finalized BSC proof header")
    if set(accounts) != set(bundle["accounts"]):
        raise VerificationError("account coverage differs from captured proofs")
    count = 0
    for selected, metadata in accounts.items():
        proven = verify_account(header["state_root"], selected, bundle["accounts"][selected]["proof"])
        if complete:
            count += verify_complete(proven, ((row["slot"], row["value"]) for row in value["storage"]
                if row["address"] == selected and int(row["value"], 16)), metadata["code"], metadata)
        else:
            verify_metadata(proven, metadata["code"], metadata)
    if not complete:
        if rpc is None:
            raise ValueError("sampled storage verification requires RPC")
        for row in value["storage"]:
            got = rpc.call("eth_getStorageAt", [row["address"], row["slot"], hex(header["number"])])
            if unhex(got, 32) != unhex(row["value"], 32):
                raise VerificationError(f"storage sample mismatch for {row['address']} slot {row['slot']}")
            count += 1
        after = rpc.call("eth_getBlockByNumber", [hex(header["number"]), False])
        if after["hash"].lower() != header["hash"]:
            raise VerificationError("RPC header changed during sampled verification")
    return {"status": "root-verified-diagnostic" if complete else "sample-parity-only",
            "header": header, "header_trust": bundle["header_trust"], "accounts": sorted(accounts),
            "storage_slots_checked": count, "storage_completeness_verified": complete,
            "account_metadata_and_code": "verified", "published_checkpoint": False}


def main(complete=False, argv=None):
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("address" if complete else "--address")
    parser.add_argument("--block", type=int, help="must equal the captured DB head; current tables have no historical reads")
    if not complete:
        parser.add_argument("--limit", type=int, default=1000)
    args = parser.parse_args(argv)
    try:
        value = snapshot(args.address, None if complete else args.limit)
        accounts = validate_snapshot(value, args.block)
        rpc = RPC()
        bundle = rpc.capture(accounts, value["header"]["number"])
        result = verify(value, bundle, complete, rpc, args.block)
        print(json.dumps(result, sort_keys=True, indent=2))
    except (ValueError, RuntimeError, OSError, KeyError, subprocess.TimeoutExpired) as error:
        parser.exit(1, f"verification failed: {error}\n")

File: crates/evm-state/src/postgres.rs
//! Coherent, fail-closed diagnostics for legacy PostgreSQL current-state tables.
//! These commands do not publish checkpoints or establish historical reads.
use crate::{
    header, process,
    proof::{address, fixed, string, unhex, verify_account, verify_complete, verify_metadata},
    rpc::RpcCall,
};
use anyhow::{ensure, Context, Result};
use serde_json::{json, Value};
use std::{
    collections::{BTreeMap, BTreeSet},
    path::Path,
    process::Command,
    time::Duration,
};

pub struct Postgres {
    dsn: String,
    options: Option<String>,
}
impl Postgres {
    pub fn new(dsn: Option<&str>, options: Option<&str>) -> Self {
        Self {
            dsn: dsn.map(str::to_owned).unwrap_or_else(|| {
                std::env::var("PG_DSN").unwrap_or_else(|_| {
                    "postgresql://dev-node:insecure-change-me-in-prod@localhost:5432/dev-node"
                        .into()
                })
            }),
            options: options.map(str::to_owned),
        }
    }
    pub fn query(&self, query: &str) -> Result<String> {
        let mut command = Command::new("psql");
        command.args([
            "-XAtq",
            "--set=ON_ERROR_STOP=1",
            "--dbname",
            &self.dsn,
            "-c",
            query,
        ]);
        if let Some(options) = &self.options {
            command.env("PGOPTIONS", options);
        }
        let output=process::capture(&mut command,Duration::from_secs(120)).map_err(|_|anyhow::anyhow!("PostgreSQL snapshot query failed; check client availability and the 16 MiB output limit"))?.context("PostgreSQL snapshot query timed out")?;
        ensure!(
            output.status.success(),
            "PostgreSQL snapshot query failed; check the database connection and schema"
        );
        String::from_utf8(output.stdout).context("invalid PostgreSQL query output")
    }
    pub fn snapshot(&self, selected: Option<&str>, limit: Option<u64>) -> Result<Value> {
        let query = snapshot_query(selected, limit)?;
        let value: Value = serde_json::from_str(&self.query(&query)?)?;
        if let Some(selected) = selected {
            let selected = address(selected)?;
            ensure!(
                value["accounts"]
                    .as_array()
                    .context("invalid database accounts")?
                    .iter()
                    .map(|r| r["address"].clone())
                    .collect::<Vec<_>>()
                    == vec![json!(selected)],
                "selected account has no complete metadata row in the database"
            );
        }
        Ok(value)
    }
}

pub fn snapshot_query(selected: Option<&str>, limit: Option<u64>) -> Result<String> {
    let selected = selected.map(address).transpose()?;
    ensure!(
        limit.is_none_or(|limit| limit > 0),
        "storage sample limit must be a positive integer"
    );
    let where_account = selected
        .as_ref()
        .map(|a| format!("WHERE a.address='{a}'"))
        .unwrap_or_default();
    let where_storage = selected
        .as_ref()
        .map(|a| format!("WHERE address='{a}'"))
        .unwrap_or_default();
    let bound = limit.map(|n| format!("LIMIT {n}")).unwrap_or_default();
    // One statement makes all three subqueries share the same MVCC snapshot.
    Ok(format!("SELECT json_build_object(
      'header',(SELECT row_to_json(h) FROM (SELECT block_num AS number,block_hash AS hash,state_root FROM blocks ORDER BY block_num DESC LIMIT 1) h),
      'accounts',(SELECT coalesce(json_agg(row_to_json(a)),'[]'::json) FROM
        (SELECT a.address,a.nonce::text AS nonce,a.balance::text AS balance,a.code_hash,'0x'||encode(c.code,'hex') AS code,
          greatest(a.block_num,a.balance_block_num,a.nonce_block_num,a.code_block_num,c.first_block_num) AS last_block
         FROM accounts a LEFT JOIN code c ON a.code_hash=c.code_hash {where_account} ORDER BY a.address) a),
      'storage',(SELECT coalesce(json_agg(row_to_json(s)),'[]'::json) FROM
        (SELECT address,slot,value,block_num AS last_block FROM storage {where_storage} ORDER BY address,slot {bound}) s))"))
}

pub fn validate_snapshot(
    value: &Value,
    requested_block: Option<u64>,
) -> Result<BTreeMap<String, Value>> {
    let header = value
        .get("header")
        .filter(|h| h.is_object())
        .context("database snapshot has no head or account metadata")?;
    let rows = value["accounts"]
        .as_array()
        .filter(|a| !a.is_empty())
        .context("database snapshot has no head or account metadata")?;
    let number = header["number"].as_u64().context("invalid database head")?;
    ensure!(
        requested_block.is_none_or(|n| n == number),
        "current tables only support their captured head; --block cannot select historical state"
    );
    fixed::<32>(string(header, "hash")?)?;
    fixed::<32>(string(header, "state_root")?)?;
    let mut accounts = BTreeMap::new();
    for row in rows {
        let selected = address(string(row, "address")?)?;
        ensure!(
            !accounts.contains_key(&selected),
            "duplicate database account"
        );
        ensure!(
            ["nonce", "balance", "code_hash", "code"]
                .iter()
                .all(|key| row.get(key).is_some_and(|v| !v.is_null())),
            "missing required account metadata or bytecode; partial state cannot pass verification"
        );
        let nonce = match &row["nonce"] {
            Value::String(text) => {
                let n = text
                    .parse::<u64>()
                    .context("database nonce exceeds uint64 or is not an integer")?;
                ensure!(
                    n.to_string() == *text,
                    "database nonce must be an exact integer"
                );
                n
            }
            Value::Number(number) => number
                .as_u64()
                .context("database nonce exceeds uint64 or is not an integer")?,
            _ => anyhow::bail!("database nonce must be an exact integer"),
        };
        fixed::<32>(string(row, "code_hash")?)?;
        unhex(string(row, "code")?)?;
        let mut row = row.clone();
        row["nonce"] = json!(nonce);
        accounts.insert(selected, row);
    }
    let storage = value["storage"]
        .as_array()
        .context("invalid database storage")?;
    for row in rows.iter().chain(storage) {
        ensure!(
            row["last_block"]
                .as_u64()
                .is_some_and(|last| last <= number),
            "database state is newer than its block marker or has invalid provenance"
        );
    }
    let mut keys = BTreeSet::new();
    for row in storage {
        let selected = address(string(row, "address")?)?;
        ensure!(
            accounts.contains_key(&selected),
            "storage account has no verified metadata"
        );
        let key = fixed::<32>(string(row, "slot")?)?;
        fixed::<32>(string(row, "value")?)?;
        ensure!(
            keys.insert((selected, key)),
            "duplicate database storage slot"
        );
    }
    Ok(accounts)
}

pub fn verify(
    value: &Value,
    bundle: &Value,
    complete: bool,
    rpc: Option<&dyn RpcCall>,
    requested_block: Option<u64>,
    work: &Path,
) -> Result<Value> {
    let accounts = validate_snapshot(value, requested_block)?;
    let header = &value["header"];
    ensure!(
        bundle["chain_id"] == 56
            && ["number", "hash", "state_root"]
                .iter()
                .all(|key| header[key] == bundle["header"][key]),
        "database head differs from the finalized BSC proof header"
    );
    if let Some(encoded) = bundle.get("header_rlp") {
        header::verify_header(
            encoded.as_str().context("invalid encoded proof header")?,
            &bundle["header"],
            None,
        )?;
    }
    ensure!(
        accounts.keys().collect::<BTreeSet<_>>()
            == bundle["accounts"]
                .as_object()
                .context("invalid captured account proofs")?
                .keys()
                .collect(),
        "account coverage differs from captured proofs"
    );
    let mut count = 0_u64;
    for (selected, metadata) in &accounts {
        let proven = verify_account(
            string(header, "state_root")?,
            selected,
            &bundle["accounts"][selected]["proof"],
        )?;
        if complete {
            std::fs::create_dir_all(work)?;
            let directory = tempfile::Builder::new()
                .prefix("postgres-root-")
                .tempdir_in(work)?;
            let slots = value["storage"]
                .as_array()
                .unwrap()
                .iter()
                .filter_map(|row| {
                    // All addresses and fixed-width words were validated above.
                    if address(row["address"].as_str().unwrap()).unwrap() != *selected
                        || fixed::<32>(row["value"].as_str().unwrap())
                            .unwrap()
                            .iter()
                            .all(|b| *b == 0)
                    {
                        None
                    } else {
                        Some(Ok((
                            row["slot"].as_str().unwrap().into(),
                            row["value"].as_str().unwrap().into(),
                        )))
                    }
                });
            count += verify_complete(
                &proven,
                slots,
                string(metadata, "code")?,
                metadata,
                &directory.path().join("storage.sqlite"),
            )?;
        } else {
            verify_metadata(&proven, string(metadata, "code")?, metadata)?;
        }
    }
    if !complete {
        let rpc = rpc.context("sampled storage verification requires RPC")?;
        let number = header["number"].as_u64().unwrap();
        for row in value["storage"].as_array().unwrap() {
            let got = rpc.call(
                "eth_getStorageAt",
                json!([row["address"], row["slot"], format!("0x{number:x}")]),
            )?;
            ensure!(
                fixed::<32>(got.as_str().context("invalid RPC storage sample")?)?
                    == fixed::<32>(string(row, "value")?)?,
                "storage sample mismatch"
            );
            count += 1;
        }
        let after = rpc.call(
            "eth_getBlockByNumber",
            json!([format!("0x{number:x}"), false]),
        )?;
        ensure!(
            string(&after, "hash")?.to_ascii_lowercase() == string(header, "hash")?,
            "RPC header changed during sampled verification"
        );
    }
    Ok(
        json!({"status":if complete {"root-verified-diagnostic"} else {"sample-parity-only"},"header":header,"header_trust":bundle["header_trust"],"accounts":accounts.keys().collect::<Vec<_>>(),"storage_slots_checked":count,"storage_completeness_verified":complete,"account_metadata_and_code":"verified","published_checkpoint":false}),
    )
}
